php/class-pwawp-app-manifest.php

```diff
@@ -0,0 +1,171 @@
+	 */
+	const MANIFEST_QUERY_ARG = 'pwawp_manifest';
+
+	/**
+	 * The theme color to use if no dynamic values are present.
+	 *
+	 * @var string
+	 */
+	const FALLBACK_THEME_COLOR = '#fff';
+
+	/**
+	 * The default manifest icon sizes.
+	 *
+	 * Copied from Jetpack_PWA_Helpers::get_default_manifest_icon_sizes().
+	 * Based on a conversation in https://github.com/GoogleChrome/lighthouse/issues/291
+	 *
+	 * @var int[]
+	 */
+	public $default_manifest_icon_sizes = array( 192, 512 );
+
+	/**
+	 * Inits the class.
+	 *
+	 * Mainly copied from Jetpack_PWA_Manifest::__construct().
+	 */
+	public function init() {
+		add_action( 'wp_head', array( $this, 'manifest_link_and_meta' ) );
+		add_action( 'amp_post_template_head', array( $this, 'manifest_link_and_meta' ) );
+		add_action( 'template_redirect', array( $this, 'send_manifest_json' ), 2 );
+	}
+
+	/**
+	 * Outputs the <link> and <meta> tags for the app manifest.
+	 *
+	 * Mainly copied from Jetpack_PWA_Manifest::render_manifest_link().
+	 */
+	public function manifest_link_and_meta() {
+		?>
+		<link rel="manifest" href="<?php echo esc_url( add_query_arg( self::MANIFEST_QUERY_ARG, '1', home_url( '/' ) ) ); ?>">
+		<meta name="theme-color" content="<?php echo esc_attr( $this->get_theme_color() ); ?>">
+		<?php
+	}
+
+	/**
+	 * Gets the theme color for the manifest.
+	 *
+	 * Mainly copied from Jetpack_PWA_Helpers::get_theme_color().
+	 * First looks for the header background color in the AMP for WordPress plugin, if it's active.
+	 * This color displays on loading the app.
+	 *
+	 * @return string $theme_color The theme color for the manifest.json file, as a hex value.
+	 */
+	public function get_theme_color() {
+		if ( class_exists( 'AMP_Customizer_Settings' ) ) {
+			/* This filter is documented in wp-content/plugins/amp/includes/class-amp-post-template.php */
+			$amp_settings = apply_filters(
+				'amp_post_template_customizer_settings',
+				AMP_Customizer_Settings::get_settings(),
+				null
+			);
+
+			if ( isset( $amp_settings['header_background_color'] ) ) {
+				$theme_color = $amp_settings['header_background_color'];
+			}
+		} elseif ( current_theme_supports( 'custom-background' ) ) {
+			$background_color = get_background_color(); // This returns a hex value without the leading #, or an empty string.
+			if ( $background_color ) {
+				$theme_color = "#$background_color";
+			}
+		}
+
+		if ( ! isset( $theme_color ) ) {
+			$theme_color = self::FALLBACK_THEME_COLOR;
+		}
+
+		/**
+		 * Enables overriding the PWA theme color.
+		 *
+		 * @param string $theme_color Hex color value.
+		 */
+		return apply_filters( 'pwawp_background_color', $theme_color );
+	}
+
+	/**
+	 * Sends the json response of the manifest.
+	 *
+	 * Mainly copied from Jetpack_PWA_Helpers::render_manifest_json().
+	 */
+	public function send_manifest_json() {
+
+		// Don't load the manifest in multiple locations.
+		if ( is_front_page() && ! empty( $_GET[ self::MANIFEST_QUERY_ARG ] ) ) { // WPCS: CSRF ok.
```

Comment:
I think we may want to register an actual query var so that we can just do `get_query_var( 'web_app_manifest' )` for example.

Reply:
On second thought, why not just leverage the REST API for the manifest endpoint?

Reply:
Sounds good. I'll probably do that tomorrow (Thursday) if that's alright.

Reply:
[3bc37](https://github.com/xwp/pwa-wp/pull/13/commits/3bc378a31ce34a6bc0621ed9e55c71ae9e4421ce) adds a REST API endpoint for the manifest, and uses it in the `<link rel="manifest">`. 

The [namespace and route](https://github.com/xwp/pwa-wp/pull/13/commits/3bc378a31ce34a6bc0621ed9e55c71ae9e4421ce#diff-2ef052ceff51a4e22091fcdab572fc69R34) still might need to be changed. Right now, the endpoint is `/wp-json/wp/v2/pwa-manifest`.

It looks like almost all of the endpoints that Core registers have the namespace of `wp/v2`:

<img width="925" alt="core-namespaces" src="https://user-images.githubusercontent.com/4063887/41738889-f93446ec-7558-11e8-82c2-5dc52b61ae23.png">